### PR TITLE
Fixes #3403

### LINKED ISSUE #3403: Doing a match of a recursive smarts leaves traces of the previous match
**Configuration:**

- RDKit Version: 2020.03
- Operating system: All
**Description:**

```
from rdkit import Chem
core1 = "[$(C-!@[a])](=O)(Cl)"
sma = Chem.MolFromSmarts(core1)

m = Chem.MolFromSmiles("c1ccccc1C(=O)Cl")
assert not m.HasSubstructMatch(sma, recursionPossible=False), "I shouldn't match (1)"

m = Chem.MolFromSmiles("c1ccccc1C(=O)Cl")
assert mh.HasSubstructMatch(sma)

m = Chem.MolFromSmiles("c1ccccc1C(=O)Cl")
assert not m.HasSubstructMatch(sma, recursionPossible=False), "I shouldn't match (2)"
```

### PR DESCRIPTION
**Pull request by bp-kelley**
Fixes #3403

### CODE REVIEW

File: Code/GraphMol/Wrap/rough_test.py [hunk 1]
```diff
@@ -6142,7 +6142,20 @@ def testMolsFromPNG(self):
     for mol, refMol in zip(mols, refMols):
       self.assertEqual(Chem.MolToSmiles(mol), Chem.MolToSmiles(refMol))
 
-
+  def test_github3403(self):
+    core1 = "[$(C-!@[a])](=O)(Cl)"
+    sma = Chem.MolFromSmarts(core1)
```
Comment (greglandrum) [inline Code/GraphMol/Wrap/rough_test.py:6147]: please use the unittest methods here instead of assert (like the rest of the tests)

**Review by greglandrum — CHANGES_REQUESTED** [review]
Aside from the inconsistent testing style this looks good

File: Code/GraphMol/Wrap/rough_test.py [hunk 2]
```diff
@@ -6142,7 +6142,20 @@ def testMolsFromPNG(self):
...
+    
+    m = Chem.MolFromSmiles("c1ccccc1C(=O)Cl")
+    assert not m.HasSubstructMatch(sma, recursionPossible=False), "I shouldn't match (1)"
+    
+    m = Chem.MolFromSmiles("c1ccccc1C(=O)Cl")
+    assert mh.HasSubstructMatch(sma)
```
Comment (greglandrum) [inline Code/GraphMol/Wrap/rough_test.py:6153]: just looked through the CI builds. `mh` isn't defined here

Comment (bp-kelley) [reply Code/GraphMol/Wrap/rough_test.py]: Whups. Fixed both.

**Review by greglandrum — APPROVED** [review]

**Pull request merged**